coral-schema/src/main/java/com/linkedin/coral/schema/avro/RelToAvroSchemaConverter.java

```diff
@@ -272,7 +272,13 @@ public RelNode visit(LogicalUnion logicalUnion) {
       RelNode relNode = super.visit(logicalUnion);
       Schema inputSchema1 = schemaMap.get(logicalUnion.getInput(0));
       Schema inputSchema2 = schemaMap.get(logicalUnion.getInput(1));
-
+      /* if forceLowercase is true, we should cast the schemas to lowercase
+       * before union to make sure column name matches
+       */
+      if (forceLowercase) {
```

Comment:
I understand that this logic is required. Although I'm not sure if this is the best place for it. Here it seems a custom transformation for merging only. On the surface, it's hard to understand "why".
 
Before reaching LogicalUnion, I think we'll always visit TableScan for the underlying individual tables. Can we visit `ToLowercaseSchemaVisitor` as part of `visit(TableScan tableScan)` instead? Then we won't need this code here

Reply:
could you please take a look at the thread I have with Jiefan? there is a test case involved and I find out this is needed

Reply:
And I think what you are suggesting here is what we are doing previously. We used to call in `getAvroSchemaForTable`, and this is being used in `getTableScanSchema`, my concern is, when we are calling `ToLowercaseSchemaVisitor`, it is casting the schema fields to lowercase, but it won't cast the final generated schema to lowercase and we have to call `ToLowercaseSchemaVisitor` after `getAvroSchemaForTable` to generated the lower cased schema. 
If we are more concerned about the readability, we can restore everything and just add
```
      if (forceLowercase) {
        schema = ToLowercaseSchemaVisitor.visit(schema);
      }
```
to the converter.

Reply:
Let's not intentionally degrade performance. 
I just noticed that `LogicalUnion` has `LogicalProject` as its child nodes. That probably means that inputSchema1 and inputSchema2 are generated by first visiting `visit(LogicalProject logicalProject)` 

Could you check that if we force lower case in `visit(LogicalProject logicalProject)` , would it be backward compatible and also resolve the issue at hand?

Else, I'm good with the above changes.

Reply:
Thanks for the suggestion!

Do you mean the function `RelNode visit(LogicalProject project)` in this class `HiveRelConverter`? 
I didn't find that we have the flag `forcelowercase` available in this class. I don't think we can move `ToLowercaseSchemaVisitor` unless the flag is accessible there.

Reply:
I just tested it by adding 
```
--- a/coral-schema/src/main/java/com/linkedin/coral/schema/avro/RelToAvroSchemaConverter.java
+++ b/coral-schema/src/main/java/com/linkedin/coral/schema/avro/RelToAvroSchemaConverter.java
@@ -214,6 +214,9 @@ public class RelToAvroSchemaConverter {
     public RelNode visit(LogicalProject logicalProject) {
       RelNode relNode = super.visit(logicalProject);
       Schema inputSchema = schemaMap.get(logicalProject.getInput());
+      if (forceLowercase) {
+        ToLowercaseSchemaVisitor.visit(inputSchema);
+      }
```
And this  can't pass the test case we just added. Is this what you are suggesting?

I think this is also caused by the same issue we are trying to solve, which is even if we cast the schema to upper case, it was overwritten to lowercase and we have to call `ToLowercaseSchemaVisitor` out side of the schema converter

Reply:
Maybe consider factoring the `forceLowercase` flag into the `SchemaUtilities.mergeUnionRecordSchema` utility method to make the method tolerant of field name case sensitivity and always return the lowercase unioned schema.

FYI @aastha25  do you think this is better?

Reply:
Thanks for the suggestion. I think that's a good middle-ground for now. It'll ensure that `SchemaUtilities .mergeUnionRecordSchema` handles casing as per merging requirements.

Reply:
thanks Raymond! fixed in the latest patch.